src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs

```diff
@@ -82,35 +83,114 @@ private async Task InvokeInternal(HttpContext context)
                 var list = new List<KeyValuePair<string, string?>>(
                     request.Headers.Count + DefaultRequestFieldsMinusHeaders);
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestProtocol))
+                if (options.ModifyRequestLog != null)
                 {
-                    AddToList(list, nameof(request.Protocol), request.Protocol);
-                }
+                    var headerDictionary = new HeaderDictionary();
+                    var loggingContext = new HttpRequestLoggingContext(context, options, headerDictionary);
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestMethod))
-                {
-                    AddToList(list, nameof(request.Method), request.Method);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestProtocol))
+                    {
+                        loggingContext.Protocol = request.Protocol;
+                    }
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestScheme))
-                {
-                    AddToList(list, nameof(request.Scheme), request.Scheme);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestMethod))
+                    {
+                        loggingContext.Method = request.Method;
+                    }
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestPath))
-                {
-                    AddToList(list, nameof(request.PathBase), request.PathBase);
-                    AddToList(list, nameof(request.Path), request.Path);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestScheme))
+                    {
+                        loggingContext.Scheme = request.Scheme;
+                    }
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestQuery))
-                {
-                    AddToList(list, nameof(request.QueryString), request.QueryString.Value);
-                }
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestPath))
+                    {
+                        loggingContext.PathBase = request.PathBase;
+                        loggingContext.Path = request.Path;
+                    }
 
-                if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestHeaders))
+                    if (options.LoggingFields.HasFlag(HttpLoggingFields.RequestQuery))
+                    {
+                        loggingContext.Query = request.QueryString.ToString();
```

Comment:
Using PathString.Value and QueryString.Value would avoid extra re-encoding, unless we know people would prefer the encoded value in the logs?

Reply:
I would have assumed people prefer the unencoded value in the logs.